Updated land perturbation scheme

ClaraDraper-NOAA opened the pull request:
1. Clean up of the pre-existing land perturbation scheme (was called with do_sfcperts) that was intended for use with GEFS forecasts. This scheme is not being used, and needed some clean-up to work robustly. This scheme has been retained as lndp_type=1 (output is zero diff to old scheme) – think about removing in the future? 

2. Addition of new land surface perturbation option (lndp_type=2), intended for use in cycling DA, and designed to sit outside of the forecast model code. 

Regressions tests have passed on hera, and I’ve added a fv3_ccpp_lndp regression test for the lndp_type=2 option. 

Detailed list of changes: 

In stochastic physics:
* logical do_sfcperts replaced with integer lndp_type (to enable selection of multiple land pert scheme options)
* deleted land_sppt namelist entry (land SPPT has not been implemented, and experiments indicate it has minimal impact)
* changed namelist names for consistent formatting, and to specify land (lnd) rather than more generic surface (sfc) 
* hard-wired npsfc (now nplnd) dimension for old scheme, as was not coded correctly for npsfc >1 (this is the number of patterns used for the surface perturbations)
* replaced hardwired index in old scheme for the perturbed variables in sfc_wts with scheme to assign indexes according to which variables are specified for perturbations
* merged run_stochastic_physics_sfc routine into run_stochastic_physics
* simplified namelist inputs, by replacing individual variables in each namelist with an input vector of variable codes for which perturbations are requested (lndp_var_list), and a vector of perturbation magnitudes (lndp_prt_list)
*For lndp_type=2,  added capacity to apply land perturbations throughout forecast (do_advance_pattern), and to include the land perturbations in the stochini file.


Also, fixed a couple of minor bugs, most notably: 
*added missing error trapping in stochastic physics (may not have done this everywhere that is necessary). 

In FV3: 
*merged run_stochastic_physics_sfc routine into run_stochastic_physics 
* added calls to new land perturbation scheme 
* added error trapping for abnormal exit from run_stochastic_physics 
* added GFS_land_perts module with routines to apply the new perturbations 
* replaced hardwired index in old scheme for the perturbed variables in sfc_wts with scheme to assign indexes according to which variables are specified for perturbations
* various changes to variables in namelists and type_defs

In ccpp/framework 
* variable name updates

In ccpp/physics: 

*variable name updates
* replaced hardwired index in old scheme for the perturbed variables in sfc_wts with scheme to assign indexes according to which variables are specified for perturbations

In UFS: 
•	Updated variable names in namelists 
•	Addition of lndp test

ASSOCIATED PULL REQUESTS: 
stochastic_physics #28 
ccpp-framework #316
ccpp-physics #480
fv3atm #148 
ufs-weather-model #176

climbfuji commented:
@ClaraDraper-NOAA thanks for submitting all these PRs. Just for your info, if you are not making changes in ccpp-framework (you have no commits), you do not need to create PRs for those repositories. Just make sure that the submodule pointer in fv3atm points to the correct hash (= top of master) when it is your turn in the commit queue. I am going to close this PR and ask you to remove the "ccpp framework" part in your descriptions (by the way, the links for ccpp-framework point to pull request 316 of ccpp-physics, not ccpp-framework.

ClaraDraper-NOAA commented:
Hi Dom,

Thanks for this. From memory there was one very small change to
ccpp-framework, so I'm guessing I messed up the PR (my first time doing
this, sorry). I'll be at the UFS User's Workshop today, but will look into
it as soon as I can (before the end of the day).

Clara


On Mon, Jul 27, 2020 at 8:27 PM Dom Heinzeller <notifications@github.com>
wrote:

> Closed #316 <https://github.com/NCAR/ccpp-framework/pull/316>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NCAR/ccpp-framework/pull/316#event-3592337920>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AH7BYD33D77VTJZS34YY3FTR5YZRTANCNFSM4PJJG3MA>
> .
>


-- 
--------------------------------------------------------------------
Clara Draper
CIRES / NOAA Earth System Research Laboratories
Physical Sciences Laboratory
325 Broadway, Boulder, CO
DSRC, Rm 1D-105
+1 303 497 4467
--------------------------------------------------------------------

ClaraDraper-NOAA commented:
Dom, 

A quick follow-up: I had earlier made a change to a file that is no longer in the ccpp-framework repository (scripts/convert_metadata.py). It appears that git is smarter than I am, as it knew I had no changes now even if I didn't! 

I've updated .git_modules to the NCAR repository, and fixed up the comments for each PR. Sorry for the confusion.
Clara
